### ISSUE TITLE
[BUG] HTTP 500 Method failed when editing a channel

### ISSUE THREAD

**Issue by N-Mi** [nextgenhealthcare/connect#5815]
**Describe the bug**
Adding, deleting, deploying, undeploying a channel works as expected.
But when trying to edit a channel, I get an error 500.

**To Reproduce**
Setup steps :
1. Right-click an existing channel and select "Clone" (didn't try by creating a channel from scratch)
2. Modify name and settings
3. Click on "Save" button
4. Channel is added and can be deployed/undeployed without issues

1. Go to "Channels"
2. Undeploy (if needed) the desired channel
3. Right-click on channel, select "Edit Channel"
4. Change any settings
5. Click on "Save" button
6. See the error

**Expected behavior**
Changes made to the channel should be saved, and then the channel should be usable with the new configuration.

**Actual behavior**
I can only only return to channel list by discarding changes.
If I need to make any changes to a channel, the only way I can do it is clone it, save, and delete the old channel.

**Screenshots**
![shot-2023-06-13_15-39-02](https://github.com/nextgenhealthcare/connect/assets/3135085/122cfb0e-92b5-4f2d-921e-a5240408fefe)

Full error message :
```
Method failed: HTTP/1.1 500 Internal Server Error
com.mirth.connect.client.core.ClientException: Method failed: HTTP/1.1 500 Internal Server Error
at com.mirth.connect.client.core.ServerConnection.handleResponse(ServerConnection.java:529)
at com.mirth.connect.client.core.ServerConnection.executeSync(ServerConnection.java:256)
at com.mirth.connect.client.core.ServerConnection.apply(ServerConnection.java:166)
at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:255)
at org.glassfish.jersey.client.JerseyInvocation$3.call(JerseyInvocation.java:722)
at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
at org.glassfish.jersey.internal.Errors.process(Errors.java:228)
at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:444)
at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:718)
at org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:459)
at org.glassfish.jersey.client.proxy.WebResourceFactory.invoke(WebResourceFactory.java:379)
at com.sun.proxy.$Proxy51.updateChannel(Unknown Source)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at com.mirth.connect.client.core.Client$2.invoke(Client.java:266)
at com.sun.proxy.$Proxy51.updateChannel(Unknown Source)
at com.mirth.connect.client.core.Client.updateChannel(Client.java:1071)
at com.mirth.connect.client.core.Client.updateChannel(Client.java:1061)
at com.mirth.connect.client.ui.Frame.updateChannel(Frame.java:1592)
at com.mirth.connect.client.ui.ChannelSetup.saveChanges(ChannelSetup.java:1253)
at com.mirth.connect.client.ui.Frame$22.doInBackground(Frame.java:2957)
at com.mirth.connect.client.ui.Frame$22.doInBackground(Frame.java:2953)
at javax.swing.SwingWorker$1.call(SwingWorker.java:295)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at javax.swing.SwingWorker.run(SwingWorker.java:334)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
at com.mirth.connect.server.controllers.DefaultChannelController.updateChannel(DefaultChannelController.java:376)
at com.mirth.connect.server.api.servlets.ChannelServlet.updateChannel(ChannelServlet.java:208)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.base/java.lang.reflect.Method.invoke(Method.java:566)
at com.mirth.connect.server.api.providers.MirthResourceInvocationHandlerProvider$1.invoke(MirthResourceInvocationHandlerProvider.java:219)
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)
at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:205)
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)
at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)
at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)
at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)
at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326)
at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)
at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)
at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
at org.glassfish.jersey.internal.Errors.process(Errors.java:267)
at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)
at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)
at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)
at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:471)
at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:425)
at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:383)
at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:336)
at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:223)
at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:864)
at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1655)
at com.mirth.connect.server.MethodFilter.doFilter(MethodFilter.java:37)
at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
at com.mirth.connect.server.api.providers.ClickjackingFilter.doFilter(ClickjackingFilter.java:44)
at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
at com.mirth.connect.server.api.providers.ApiOriginFilter.doFilter(ApiOriginFilter.java:71)
at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1642)
at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:533)
at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255)
at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203)
at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564)
at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201)
at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144)
at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:61)
at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
at org.eclipse.jetty.server.Server.handle(Server.java:531)
at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:352)
at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:260)
at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:281)
at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:291)
at org.eclipse.jetty.io.ssl.SslConnection$3.succeeded(SslConnection.java:151)
at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:102)
at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:118)
at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:319)
at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:175)
at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:133)
at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:754)
at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:672)
at java.base/java.lang.Thread.run(Thread.java:829)
```

**Environment (please complete the following information):**
 - OS: Debian 11.7
 - Java: OpenJDK 64-Bit Server VM 11.0.18
 - Connect Version : 3.8.1

**Workaround(s)**
Are there one or more workarounds for this issue currently?
No

**Additional context**
I saw that the recommended additional settings for Java 9+ were not added to mcserver.vmoptions and mcservice.vmoptions. I added them, restarted MirthConnect on server side, then client side, re-logged, and nothing changed.

I see nothing in logs. Tried to set log level to DEBUG in log4j.properties, but I still don't have anything logged when the issue happens.

**Comment by jonbartels** [comment]
`Caused by: java.lang.NullPointerException
at com.mirth.connect.server.controllers.DefaultChannelController.updateChannel(DefaultChannelController.java:376)
at com.mirth.connect.server.api.servlets.ChannelServlet.updateChannel(ChannelServlet.java:208)`

What is the DefaultChannelController doing on line 376 in MC 3.8.1?

**Comment by N-Mi** [comment]
```
   339	    public synchronized boolean updateChannel(Channel channel, ServerEventContext context, boolean override) throws ControllerException {
   340	        // Extract metadata and then clear it from the channel model so it won't be stored in the database
   341	        ChannelExportData exportData = channel.getExportData();
   342	        channel.clearExportData();
   343	
   344	        /*
   345	         * Methods that update the channel must be synchronized to ensure the channel cache and
   346	         * database never contain different versions of a channel.
   347	         */
   348	
   349	        int newRevision = channel.getRevision();
   350	        int currentRevision = 0;
   351	
   352	        Channel matchingChannel = getChannelById(channel.getId());
   353	
   354	        // If the channel exists, set the currentRevision
   355	        if (matchingChannel != null) {
   356	            /*
   357	             * If the channel in the database is the same as what's being passed in, don't bother
   358	             * saving it.
   359	             * 
   360	             * Ignore the channel revision and last modified date when comparing the channel being
   361	             * passed in to the existing channel in the database. This will prevent the channel from
   362	             * being saved if the only thing that changed was the revision and/or the last modified
   363	             * date. The client/CLI take care of this by passing in the proper revision number, but
   364	             * the API alone does not.
   365	             */
   366	            if (EqualsBuilder.reflectionEquals(channel, matchingChannel, new String[] {
   367	                    "lastModified", "revision" })) {
   368	                updateChannelMetadata(channel.getId(), exportData.getMetadata());
   369	                updateChannelTags(channel.getId(), exportData.getChannelTags());
   370	                return true;
   371	            }
   372	
   373	            currentRevision = matchingChannel.getRevision();
   374	
   375	            // Use the larger nextMetaDataId to ensure a metadata ID will never be reused if an older version of a channel is imported or saved.
   376	            channel.setNextMetaDataId(Math.max(matchingChannel.getNextMetaDataId(), channel.getNextMetaDataId()));
   377	        }
```

**Comment by jonbartels** [comment]
A null pointer exception from line 376 should not be possible. https://github.com/nextgenhealthcare/connect/blob/c0350a872b8b2a64145f742355c3cb6e29f2cd80/server/src/com/mirth/connect/server/controllers/DefaultChannelController.java#L376

`channel.setNextMetaDataId(Math.max(matchingChannel.getNextMetaDataId(), channel.getNextMetaDataId()));`

`channel` cannot be null because other operations have already been done on it.
`matchingChannel` cannot be null because this code is inside an if condition that requires it to be not-null.
`getNextMetaDataId()` cannot be null because the `Channel` class never lets the value be null.

**Comment by N-Mi** [comment]
Maybe I see something : 
```
    public Channel() {
        destinationConnectors = new ArrayList<Connector>();
        properties = new ChannelProperties();
        nextMetaDataId = 1;
    }

    public Channel(String id) {
        this.id = id;
    }
```
The first constructor initializes `nextMetaDataId` to 1, but the second doesn't.
Maybe the second constructor is called when creating a channel using the "Clone channel" option ?

**Comment by jonbartels** [comment]
Have there been any recent changes to the MC config? Especially `vmoptions` or `properties` files?

What database are you using?

**Comment by N-Mi** [comment]
The issue is present since a at least a few months.  I'm new in the team, and the guy who handed me the service maintenance was already unable to edit a channel (he's no longer in the team btw, so I can't ask him when the issue appeared).

Before I edited vmoptions to add settings for Java9+, the file modification was not recent (but can't find the date).
Here is its content :
```
-server
-Xmx256m
-Djava.awt.headless=true
-Dapple.awt.UIElement=true
--add-modules=java.sql.rowset
--add-exports=java.base/com.sun.crypto.provider=ALL-UNNAMED
--add-exports=java.base/sun.security.provider=ALL-UNNAMED
--add-opens=java.base/java.lang=ALL-UNNAMED
--add-opens=java.base/java.lang.reflect=ALL-UNNAMED
--add-opens=java.base/java.math=ALL-UNNAMED
--add-opens=java.base/java.net=ALL-UNNAMED
--add-opens=java.base/java.security=ALL-UNNAMED
--add-opens=java.base/java.security.cert=ALL-UNNAMED
--add-opens=java.base/java.text=ALL-UNNAMED
--add-opens=java.base/java.util=ALL-UNNAMED
--add-opens=java.base/sun.security.pkcs=ALL-UNNAMED
--add-opens=java.base/sun.security.rsa=ALL-UNNAMED
--add-opens=java.base/sun.security.x509=ALL-UNNAMED
--add-opens=java.desktop/java.awt=ALL-UNNAMED
--add-opens=java.desktop/java.awt.font=ALL-UNNAMED
--add-opens=java.xml/com.sun.org.apache.xalan.internal.xsltc.trax=ALL-UNNAMED

```

Database is MariaDB 1:10.5.19-0+deb11u2

**Comment by pacmano1** [comment]
Is this just one channel or all?

**Comment by N-Mi** [comment]
Here is content of mirth.properties

```
# Mirth Connect configuration file

# directories
dir.appdata = appdata
dir.tempdata = ${dir.appdata}/temp

# ports
http.port = 8080
https.port = 8443

# password requirements
password.minlength = 0
password.minupper = 0
password.minlower = 0
password.minnumeric = 0
password.minspecial = 0
password.retrylimit = 0
password.lockoutperiod = 0
password.expiration = 0
password.graceperiod = 0
password.reuseperiod = 0
password.reuselimit = 0

# Only used for migration purposes, do not modify
version = 3.8.1

# keystore
keystore.path = ${dir.appdata}/keystore.jks
keystore.storepass = XXXXXXXXXXXXXXXXX
keystore.keypass = XXXXXXXXXXXXXXXXXX
keystore.type = JCEKS

# server
http.contextpath = /
server.url =

http.host = 0.0.0.0
https.host = 0.0.0.0

https.client.protocols = TLSv1.3,TLSv1.2,TLSv1.1
https.server.protocols = TLSv1.3,TLSv1.2,TLSv1.1,SSLv2Hello
https.ciphersuites = TLS_CHACHA20_POLY1305_SHA256,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,TLS_DHE_RSA_WITH_CHACHA20_POLY1305_SHA256,TLS_AES_256_GCM_SHA384,TLS_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDH_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDH_RSA_WITH_AES_256_GCM_SHA384,TLS_DHE_RSA_WITH_AES_256_GCM_SHA384,TLS_DHE_DSS_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDH_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDH_RSA_WITH_AES_128_GCM_SHA256,TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,TLS_DHE_DSS_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384,TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384,TLS_RSA_WITH_AES_256_CBC_SHA256,TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA384,TLS_ECDH_RSA_WITH_AES_256_CBC_SHA384,TLS_DHE_RSA_WITH_AES_256_CBC_SHA256,TLS_DHE_DSS_WITH_AES_256_CBC_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,TLS_RSA_WITH_AES_256_CBC_SHA,TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA,TLS_ECDH_RSA_WITH_AES_256_CBC_SHA,TLS_DHE_RSA_WITH_AES_256_CBC_SHA,TLS_DHE_DSS_WITH_AES_256_CBC_SHA,TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256,TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256,TLS_RSA_WITH_AES_128_CBC_SHA256,TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA256,TLS_ECDH_RSA_WITH_AES_128_CBC_SHA256,TLS_DHE_RSA_WITH_AES_128_CBC_SHA256,TLS_DHE_DSS_WITH_AES_128_CBC_SHA256,TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,TLS_RSA_WITH_AES_128_CBC_SHA,TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA,TLS_ECDH_RSA_WITH_AES_128_CBC_SHA,TLS_DHE_RSA_WITH_AES_128_CBC_SHA,TLS_DHE_DSS_WITH_AES_128_CBC_SHA,TLS_EMPTY_RENEGOTIATION_INFO_SCSV
https.ephemeraldhkeysize = 2048

# CORS headers
server.api.accesscontrolalloworigin = *
server.api.accesscontrolallowcredentials = false
server.api.accesscontrolallowmethods = GET, POST, DELETE, PUT
server.api.accesscontrolallowheaders = Content-Type
server.api.accesscontrolexposeheaders =
server.api.accesscontrolmaxage =

# Determines whether or not channels are deployed on server startup.
server.startupdeploy = true

# Determines whether libraries in the custom-lib directory will be included on the server classpath.
# To reduce potential classpath conflicts you should create Resources and use them on specific channels/connectors instead, and then set this value to false.
server.includecustomlib = false

# administrator
administrator.maxheapsize = 512m

# properties file that will store the configuration map and be loaded during server startup
configurationmap.path = ${dir.appdata}/configuration.properties

# The language version for the Rhino JavaScript engine (supported values: 1.0, 1.1, ..., 1.8, es6).
rhino.languageversion = es6

# options: derby, mysql, postgres, oracle, sqlserver
database = mysql

# examples:
#	Derby		jdbc:derby:${dir.appdata}/mirthdb;create=true
#	PostgreSQL	jdbc:postgresql://localhost:5432/mirthdb
# 	MySQL		jdbc:mysql://localhost:3306/mirthdb
#	Oracle		jdbc:oracle:thin:@localhost:1521:DB
#	SQLServer	jdbc:jtds:sqlserver://localhost:1433/mirthdb
database.url = jdbc:mysql://localhost:3306/mirthdb?serverTimezone=UTC

# if using a custom driver, specify it here
#database.driver = 

# Maximum number of connections allowed for the main read/write connection pool
database.max-connections = 20
# Maximum number of connections allowed for the read-only connection pool
database-readonly.max-connections = 20

# database credentials
database.username = mirthuser
database.password = XXXXXXXXXXXX

# If true, various read-only statements are separated into their own connection pool.
# By default the read-only pool will use the same connection information as the master pool,
# but you can change this with the "database-readonly" options. For example, to point the
# read-only pool to a different JDBC URL:
#
# database-readonly.url = jdbc:...
# 
database.enable-read-write-split = true
```

> Is this just one channel or all?

All channels I have tried to edit have the issue.
Most of the time,  I don't touch existing channels and only need to add a new channel, so I clone an existing channel, change a few settings and saves. Sometimes, there is a mistake in the configuration of the new channel, so I need to go back to the newly created channel to edit it (and end up re-cloning it to be able to apply changes).

**Comment by pacmano1** [comment]
Can you export and import a channel to a new instance of mirth and see the same behavior?

**Comment by N-Mi** [comment]
Do you mean : export a channel and import to another server running Mirth ? I only have one server running Mirth.

**Comment by N-Mi** [comment]
I just did an export of a faulty channel.
Then did an import of the xml file (on the same server). When asked if I wanted to overwrite, I choose to create a new channel.
Then modified  the channel description, clicked on 'Save' : OK
Then modified again the channel description, clicked on 'Save' : error 500

**Comment by pacmano1** [comment]
Yes.  I am then asking you to stand up another instance of Mirth (e.g. on your workstation) and try to export a channel and import it.   It can just use Derby, you don't need to replicate the complete set up - just trying to understand if there is channel corruption.   

- If you create a new channel from scratch on the current engine and clone it, can you edit it without a problem?  
- Does a channel export cleanly? 
- Can you import a full export of the full config on the same engine you will set up for the first paragraph?

**Comment by pacmano1** [comment]
You sure mirth has the correct permissions on the maria db?

**Comment by N-Mi** [comment]
Here are current permissions : 
```
MariaDB [(none)]> show grants for mirthuser@localhost;
+------------------------------------------------------------------------------------------------------------------+
| Grants for mirthuser@localhost                                                                                   |
+------------------------------------------------------------------------------------------------------------------+
| GRANT USAGE ON *.* TO `mirthuser`@`localhost` IDENTIFIED BY PASSWORD '*XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX' |
| GRANT ALL PRIVILEGES ON `mirthdb`.* TO `mirthuser`@`localhost` WITH GRANT OPTION                                 |
| GRANT ALL PRIVILEGES ON `mirthdatabase`.* TO `mirthuser`@`localhost` WITH GRANT OPTION                           |
+------------------------------------------------------------------------------------------------------------------+
3 rows in set (0,000 sec)
```

**Comment by pacmano1** [comment]
I would try this on another "clean install" engine - also I assume you checked the server log on startup to make sure that the correct JVM was starting if you have more than one installed and that no one has updated that JVM with conflicting libraries.

**Comment by N-Mi** [comment]
OK, so I just downloaded a 3.8.1 tarball on my laptop (running Debian Unstable).
Launched mcservice, and  launched same "Mirth Connect Administrator Launcher" as the production one, but connected to localhost:8443 instead.
Connected as admin/admin, and got an empty Mirth instance.
Imported the channel I previously exported from production, saved, OK
Tried to edit the channel, save, Error 500 (the same)

**Comment by pacmano1** [comment]
very interesting.   and on your new instance, does a new channel do that?

**Comment by N-Mi** [comment]
I'm opening the channel export file to check if I can clean sensitive data (passwords, hospital name, ...)

**Comment by N-Mi** [comment]
> very interesting. and on your new instance, does a new channel do that?

Seems to work fine : 
 - created a channel "foo" with description "bar", saved, OK
 - edit channel "foo", add "baz" to the description , save, OK

**Comment by N-Mi** [comment]
Cloning the "foo" channel works too, and editing the clone also.

**Comment by N-Mi** [comment]
Here is a cleaned export file that can reproduce the error : 

[export.xml.txt](https://github.com/nextgenhealthcare/connect/files/11737134/export.xml.txt)

I see there are scripts in it, maybe they are the culprit.

**Comment by pacmano1** [comment]
As exported and shared here your message templates are garbage:

![image](https://github.com/nextgenhealthcare/connect/assets/44065187/b4873efd-e719-4a60-9497-a7e9e780c22a)

**Comment by pacmano1** [comment]
If you did that after export you need real base64 stuff there or empty.

**Comment by N-Mi** [comment]
Is this better ?

[export.xml.txt](https://github.com/nextgenhealthcare/connect/files/11737401/export.xml.txt)

**Comment by pacmano1** [comment]
I see what you see and never seen that before.  As you can imagine lots of people would be complaining if this was a bug for others and not just your instance.

**Comment by N-Mi** [comment]
Do you have an idea of what is causing the issue in this channel configuration ?

**Comment by pacmano1** [comment]
I don't. I suspect your channel XML is corrupt.   

I think I did this but was spending too much time on your problem so had to stop:

- export a channel
- import to newer mirth
- save but do not edit 
- export again from newer mirth
- import that export 

I think I was able to edit the channel normally after that but I haven't confirmed that twice.  

What you can try on your own is on clean install of Mirth using your current version:

- manually build the exact same channel
- export the channel
- diff that channel with one from the bad engine and see what differences there are, note that mirth is not good about xml node order when you to the diff.  

Also do a detailed compare of the installs such that you know for a fact someone didn't override a core mirth library in an effort to patch some CVE.  

Lastly you have  characters like an "ā"  (not that one exactly, I mean characters not used in English) in comments, I doubt that's a problem of course since xml should support them but you might strip them out.  

Lastly, what is the native encoding you have for Maria DB?

**Comment by N-Mi** [comment]
> Lastly, what is the native encoding you have for Maria DB?

```
MariaDB [(none)]> SHOW VARIABLES LIKE 'char%';
+--------------------------+----------------------------+
| Variable_name            | Value                      |
+--------------------------+----------------------------+
| character_set_client     | utf8                       |
| character_set_connection | utf8                       |
| character_set_database   | utf8mb4                    |
| character_set_filesystem | binary                     |
| character_set_results    | utf8                       |
| character_set_server     | utf8mb4                    |
| character_set_system     | utf8                       |
| character_sets_dir       | /usr/share/mysql/charsets/ |
```

**Comment by pacmano1** [comment]
Like I said, not sure at this point.  I don't work  for Nextgen, just trying to help out another user.  Maybe someone else will chime in with something obvious.

**Comment by N-Mi** [comment]
Thanks for your help, I will try the procedure you described and continue investigating.

**Comment by N-Mi** [comment]
Seems I can't import in 3.8.1 a channel exported from 4.3.0.
@pacmano1 What newer version did you use ?

**Comment by pacmano1** [comment]
4.2.   A container.  The default one from Nextgen.   imported the one you posted here.   So not an apples to apples comparison.

**Comment by N-Mi** [comment]
I managed to fix my export on my local instance running 3.8.1 :partying_face: 

Following our reflexion on what could be null in the channel object : 

> `channel` cannot be null because other operations have already been done on it.
> `matchingChannel` cannot be null because this code is inside an if condition that requires it to be not-null.
> `getNextMetaDataId()` cannot be null because the `Channel` class never lets the value be null.

I decided to investigate on the last one which was the only one initialized outside the function.

I created a new almost-empty channel on my local instance, exported it and compared the two xml files I got.

Here is what I found :
```
$ grep -i metadataid XXXXXXXXXXXXXXXX.xml 
    <metaDataId>0</metaDataId>
      <metaDataId>3</metaDataId>
      <metaDataId>2</metaDataId>
$ grep -i metadataid test-new.xml 
  <nextMetaDataId>2</nextMetaDataId>
    <metaDataId>0</metaDataId>
      <metaDataId>1</metaDataId>
```

Two obvious differences : 
 - no ` nextMetaDataId` attribute
 - metadatas not ordered (I think it's probably not important) and not in sequence

So I renumbered `metaDataId` 3 to 2, and added a `nextMetaDataId` attribute : 
```
$ grep -i metadataid XXXXXXXXXXXXXXXX.xml 
  <nextMetaDataId>3</nextMetaDataId>
    <metaDataId>0</metaDataId>
      <metaDataId>1</metaDataId>
      <metaDataId>2</metaDataId>
```

Imported the modified xml file to my local instance, saved, OK.
Edited the channel again, and now I can save my modifications without error.

Don't know the reason for this faulty config, but it seems fixed now. Now I just have to replicate this on my production instance.
